ruby/lib/goi/nihongo/conjugator.rb

```diff
@@ -0,0 +1,346 @@
+
+      end
+
+      class PatternRule
+        def initialize(pattern, replacement)
+          @pattern = pattern
+          @replacement = replacement
+        end
+
+        attr_reader :pattern
+        attr_reader :replacement
+
+        def matches?(dictionary_spelling)
+          pattern =~ dictionary_spelling
+        end
+
+        def call(dictionary_spelling)
+          # TODO: Should be able to track when a pattern doesn't match and give an error response.
+          dictionary_spelling.sub(pattern, replacement)
+        end
+      end
+
+      def self.concat_sets(a, b)
+        h = Hash.new
+        ks = (a.keys + b.keys).uniq
+        ks.each do |k|
+          h[k] = if a.has_key?(k) && b.has_key?(k)
+                   SetRule.new(a[k], b[k])
+                 else
+                   a[k] || b[k]
+                 end
+        end
+        h
+      end
+
+      # Type safe convenience for the standard set of verb conjugations
+      def self.verb_set(positive_plain_present: nil,
+                        positive_plain_past: nil,
+                        positive_plain_te: nil,
+                        negative_plain_present: nil,
+                        negative_plain_past: nil,
+                        negative_plain_te: nil,
+                        positive_polite_present: nil,
+                        positive_polite_past: nil,
+                        negative_polite_present: nil,
+                        negative_polite_past: nil)
+        {
+          inflection('POSITIVE', 'PLAIN', 'PRESENT') => positive_plain_present,
+          inflection('POSITIVE', 'PLAIN', 'PAST') => positive_plain_past,
+          inflection('POSITIVE', 'PLAIN', 'TE') => positive_plain_te,
+
+          inflection('NEGATIVE', 'PLAIN', 'PRESENT') => negative_plain_present,
+          inflection('NEGATIVE', 'PLAIN', 'PAST') => negative_plain_past,
+          inflection('NEGATIVE', 'PLAIN', 'TE') => negative_plain_te,
+
+          inflection('POSITIVE', 'POLITE', 'PRESENT') => positive_polite_present,
+          inflection('POSITIVE', 'POLITE', 'PAST') => positive_polite_past,
+
+          inflection('NEGATIVE', 'POLITE', 'PRESENT') => negative_polite_present,
+          inflection('NEGATIVE', 'POLITE', 'PAST') => negative_polite_past,
+        }.compact
+      end
+
+      # Type safe convenience for the standard set of adjective conjugations
+      def self.adjective_set(positive_plain_present: nil,
+                             positive_plain_past: nil,
+                             negative_plain_present: nil,
+                             negative_plain_past: nil)
+        {
+          inflection('POSITIVE', 'PLAIN', 'PRESENT') => positive_plain_present,
+          inflection('POSITIVE', 'PLAIN', 'PAST') => positive_plain_past,
+
+          inflection('NEGATIVE', 'PLAIN', 'PRESENT') => negative_plain_present,
+          inflection('NEGATIVE', 'PLAIN', 'PAST') => negative_plain_past
+        }.compact
+      end
+
+      def self.godan_rule_set(
```

Comment:
I can't find examples of a few of these endings, but in many cases there is a clear pattern so I just let it be, and in others I explicitly have no matching rule so that we error the first time we see one.